File: all/1653_minimum_deletions_to_make_string_balanced/sol_dp_better_space.c
static inline int min(const int i1, const int i2) {
    return ((i1 < i2) ? i1 : i2);
}

int minimumDeletions(char* s) {
    int current_dp_state = 0;
    int previous_total_b_count = 0;

    for (int i = 0; '\0' != s[i]; ++i) {
        if ('b' == s[i]) {
            ++previous_total_b_count;
        }
        else if ('a' == s[i]) {
            current_dp_state = min(
                current_dp_state + 1,
                previous_total_b_count
            );
        }
    }

    return current_dp_state;
}

File: all/1653_minimum_deletions_to_make_string_balanced/sol_dp_better_space.py
class Solution:
    def minimumDeletions(self, s: str) -> int:
        current_dp_state = 0
        previous_total_b_count = 0

        for index, c in enumerate(s):
            if 'b' == c:
                previous_total_b_count += 1
            elif 'a' == c:
                current_dp_state = min(
                    current_dp_state + 1,
                    previous_total_b_count
                )

        return current_dp_state


if '__main__' == __name__:
    sol = Solution()
    testcases_list = [
        ['aababbab'],
        ['bbaaaaabb'],
    ]

    for testcase in testcases_list:
        print(sol.minimumDeletions(*testcase))
